mkosi/__init__.py

```diff
@@ -294,45 +294,50 @@ def mount_build_overlay(state: MkosiState, read_only: bool = False) -> ContextMa
     return mount_overlay([state.root], state.workspace / "build-overlay", state.root, read_only)
 
 
+def finalize_sources(config: MkosiConfig) -> list[tuple[Path, Path]]:
+    sources = [
+        (src, Path("work/src") / (str(target).lstrip("/") if target else "."))
+        for src, target
+        in config.build_sources
+    ]
+
+    return sorted(sources, key=lambda s: s[1])
+
+
 def run_prepare_script(state: MkosiState, build: bool) -> None:
     if state.config.prepare_script is None:
         return
     if build and state.config.build_script is None:
         return
 
     bwrap: list[PathString] = [
-        "--bind", state.config.build_sources, "/root/src",
-        "--bind", state.config.prepare_script, "/root/prepare",
-        "--chdir", "/root/src",
+        "--bind", state.config.prepare_script, "/work/prepare",
+        "--chdir", "/work/src",
     ]
 
-    def clean() -> None:
-        srcdir = state.root / "root/src"
-        if srcdir.exists():
-            srcdir.rmdir()
-
-        state.root.joinpath("root/prepare").unlink()
+    for src, target in finalize_sources(state.config):
+        bwrap += ["--bind", src, Path("/") / target]
 
     if build:
         with complete_step("Running prepare script in build overlay…"), mount_build_overlay(state):
             run_workspace_command(
                 state.root,
-                ["/root/prepare", "build"],
+                ["/work/prepare", "build"],
                 network=True,
                 bwrap_params=bwrap,
-                env=dict(SRCDIR="/root/src") | state.environment,
+                env=dict(SRCDIR="/work/src") | state.environment,
```

Comment:
Shouldn't that actually be the other way 'round? This way `SRCDIR` in `state.environment` can clobber our `SRCDIR`.

Reply:
Yeah, but we do this in a few places, let's address this in another PR